### ISSUE TITLE
Should not use ncid from different libraries

### ISSUE THREAD

**Issue by edhartnett** [NCAR/ParallelIO#80]
Currently the code uses the ncid to find the netCDF file. But the libraries below assign ncid in different ways:

netcdf-3.x starts with 4.
pnetcdf starts with 0
netcdf-4 starts with 2^16

So there will be no problem until a user with netcdf-3.x has one file open in classic (serial) mode, and opens a 5th file with pnetcdf. Then the ncids will overlap and PIO will get confused.

It was possible to build netcdf-3.x, even well into the netCDF-4 era, but using --disable-netcdf-4 in configure, so it is possible that there are still HPC installations of netcdf-3.x libraries.

Jim indicated recently that he believes it is a reasonable scenario that a user could have some files open in netcdf serial mode, and several more files open with pnetcdf. So this should probably be fixed.

The fix is easy, just maintain an internal array of PIO ncids, which will be passed out to users from the openfile and createfile functions. Then pio_get_file_from_id() is changed to use the internal ncid, and ensure that all function calls to the underlying libraries are changed to use the ncid from file-fh.

**Comment by jayeshkrishna** [comment]
I agree.
Another related problem that I have found in my testing is that PIO assumes ncids returned from all the io processes are the same. This is not guaranteed (atleast in pnetcdf). Consider the case where an MPI process opens different files using different communicators (iosystems). The pnetcdf developers suggest that we consider ncid as an opaque integer.
One way to fix the issue (that I have used locally for my testing with ACME) above would be to modify pio_get_file_from_id() to also accept the iosystem id. 
I am working on a test to recreate the problem and will create a new bug report for it.

**Comment by edhartnett** [comment]
OK, good to know about pnetcdf and their ncids. A little bit tricky on their part.

But that may break async feature with pnetcdf. The compute task sends the ncid for a PIOc_close to the IO master task, which uses it to close the file.

So this fix of the ncids is actually crucial to getting async working with the pnetcdf library. Thanks for making that clear jayeshkrishna!

**Comment by edhartnett** [comment]
Fixed by recent merge.

**Comment by rljacob** [comment]
Which PR fixed this?  Put  "Fixes #80" in the PR description and it will autoclose this issue.

**Comment by edhartnett** [comment]
OK, I didn't know that about put "fixes" in the PR description. I'll remember that for the future!

This problem was solved by the new ncid code we just merged. Since ncids are now assigned internally by PIO, the ncids that are provided by the base libraries (netCDF, pnetcdf, and netCDF-4) no longer get exposed to the user.

As a result it does not matter if the internal ncids overlap. The external ncids (the ones assigned by PIO) will never overlap.
